File: Agents/Azure-Agent-Skills-In-Action/skill-demos/sdk-typescript/04_azure_cosmos.ts
// TRIPLE:
//   Skill: azure-cosmos-ts
//   Prompt: "Using azure-cosmos-ts skill, write TypeScript code that connects to Cosmos DB, creates a database/container, upserts an item, and queries by partition key."
//   Deliverable: This file
//   Source: https://github.com/microsoft/skills/tree/main/.github/plugins/azure-sdk-typescript/skills/azure-cosmos-ts

import { CosmosClient } from "@azure/cosmos";
import { DefaultAzureCredential } from "@azure/identity";

async function main(): Promise<void> {
  const endpoint = process.env["AZURE_COSMOS_ENDPOINT"]!;
  const credential = new DefaultAzureCredential();
  const client = new CosmosClient({ endpoint, aadCredentials: credential });

  // Create database if not exists
  const { database } = await client.databases.createIfNotExists({ id: "skill-demo-db" });
  console.log(`Database: ${database.id}`);

  // Create container with partition key
  const { container } = await database.containers.createIfNotExists({
    id: "items",
    partitionKey: { paths: ["/category"] },
  });
  console.log(`Container: ${container.id}`);

  // Upsert an item
  const item = {
    id: "item-001",
    category: "demo",
    name: "Azure Cosmos Skill Demo",
    timestamp: new Date().toISOString(),
  };
  const { resource: upserted } = await container.items.upsert(item);
  console.log(`Upserted: ${upserted?.id}`);

  // Query by partition key
  const querySpec = {
    query: "SELECT * FROM c WHERE c.category = @category",
    parameters: [{ name: "@category", value: "demo" }],
  };
  const { resources } = await container.items.query(querySpec).fetchAll();
  console.log(`Query results: ${resources.length} item(s)`);
  for (const r of resources) {
    console.log(`  ${r.id}: ${r.name}`);
  }
}

main().catch(console.error);

File: Agents/Azure-Agent-Skills-In-Action/skill-demos/sdk-java/04_AzureCosmos.java
// TRIPLE:
//   Skill: azure-cosmos-java
//   Prompt: "Using azure-cosmos-java skill, write Java code that creates a Cosmos DB database/container, upserts an item, and queries."
//   Deliverable: This file
//   Source: https://github.com/microsoft/skills/tree/main/.github/plugins/azure-sdk-java/skills/azure-cosmos-java

import com.azure.cosmos.CosmosClient;
import com.azure.cosmos.CosmosClientBuilder;
import com.azure.cosmos.CosmosContainer;
import com.azure.cosmos.CosmosDatabase;
import com.azure.cosmos.models.*;
import com.azure.identity.DefaultAzureCredentialBuilder;

import java.time.Instant;
import java.util.Map;

public class AzureCosmos {

    public static void main(String[] args) {
        String endpoint = System.getenv("AZURE_COSMOS_ENDPOINT");

        CosmosClient client = new CosmosClientBuilder()
                .endpoint(endpoint)
                .credential(new DefaultAzureCredentialBuilder().build())
                .consistencyLevel(ConsistencyLevel.SESSION)
                .buildClient();

        // Create database if not exists
        CosmosDatabaseResponse dbResp = client.createDatabaseIfNotExists("skill-demo-db");
        CosmosDatabase database = client.getDatabase(dbResp.getProperties().getId());
        System.out.printf("Database: %s%n", database.getId());

        // Create container with partition key
        CosmosContainerProperties containerProps =
                new CosmosContainerProperties("items", "/category");
        database.createContainerIfNotExists(containerProps);
        CosmosContainer container = database.getContainer("items");
        System.out.printf("Container: %s%n", container.getId());

        // Upsert an item
        Map<String, Object> item = Map.of(
                "id", "item-001",
                "category", "demo",
                "name", "Azure Cosmos Skill Demo",
                "timestamp", Instant.now().toString()
        );
        container.upsertItem(item);
        System.out.println("Upserted: item-001");

        // Query by partition key
        String query = "SELECT * FROM c WHERE c.category = 'demo'";
        CosmosQueryRequestOptions options = new CosmosQueryRequestOptions();
        container.queryItems(query, options, Map.class)
                .forEach(doc -> System.out.printf("  %s: %s%n", doc.get("id"), doc.get("name")));

        client.close();
    }
}
